### PR TITLE
Add maxWorkers option

### LINKED ISSUE #2022: Add an option to set the number of spawned WebWorkers
I am working on the app where we need to be able to control / reduce the number of web workers that mapbox-gl uses. Looks like now it is always _navigator.hardwareConcurrency - 1_ ?  Is there a reason why the number of workers should not be exposed / overwritten via API?

### PR DESCRIPTION
**Pull request by anandthakker**
Also sets maxWorkers on the debug page to 3, on the theory that a pretty
large majority of users have 4 cpus.

Closes #2022

### CODE REVIEW

**Comment by anandthakker** [comment]
cc @lucaswoj 

I don't see an especially good way to add meaningful tests for this, since it's basically about passing a number through to `new Style()` and then to `new Dispatcher()`.  I've used something like `proxyquire` in other projects to mock CJS module dependencies, but not sure that that's something yall want to adpot

**Comment by lucaswoj** [comment]
> I don't see an especially good way to add meaningful tests for this

Looks like you made [one test](https://github.com/mapbox/mapbox-gl-js/blob/master/test/js/style/style.test.js#L1125) start failing. That might be a good place to start looking. 😉

File: debug/index.html [hunk 1]
```diff
@@ -58,7 +58,8 @@
     zoom: 12.5,
     center: [-77.01866, 38.888],
     style: 'mapbox://styles/mapbox/streets-v8',
-    hash: true
+    hash: true,
+    maxWorkers: 3
```
Comment (lucaswoj) [inline debug/index.html]: Why **`max`**`Workers`? Isn't this literally the number of workers?

**Comment by lucaswoj** [comment]
Happy to 🚢  this with a test case, and agreenment on [the parameter name](https://github.com/mapbox/mapbox-gl-js/pull/2655#discussion_r65629384).

File: debug/index.html (hunk 1 above)
Comment (mourner) [inline debug/index.html]: 👍 `numWorkers`

**Comment by anandthakker** [comment]
> Looks like you made one test start failing. That might be a good place to start looking. 😉

Oh dear. 😊 

Thanks, will do.

File: debug/index.html (hunk 1 above)
Comment (anandthakker) [inline debug/index.html]: @lucaswoj @mourner I considered "num", but chose "max" -- this was the argument in my head:

`maxWorkers` is really what the user (me) wants to control, whereas `numWorkers` is more of an internal property.  The fact that they are the same right now is an implementation detail; I could imagine the implementation changing to, say, construct/destroy workers on demand (especially given the [possibility of worker pooling](https://github.com/mapbox/mapbox-gl-js/issues/899)).  

That said, I don't feel especially strongly about this, so just lemme know what yall think

**Comment by anandthakker** [comment]
@lucaswoj Fixed the failing test and added another, but the build still fails because coveralls (I think) doesn't like my fork:
```
/home/ubuntu/mapbox-gl-js/node_modules/coveralls/bin/coveralls.js:18
        throw err;
        ^
Bad response: 422 {"message":"Couldn't find a repository matching this job.","error":true}
```

**Comment by lucaswoj** [comment]
The coveralls error should have been (might have been?) fixed by https://github.com/mapbox/mapbox-gl-js/commit/1b25043da7b3bd1c2e431d2db7ff8ff25e9241ca. Have you rebased in the last day?

**Comment by anandthakker** [comment]
> The coveralls error should have been (might have been?) fixed by 1b25043. Have you rebased in the last day?

I had not!  Just rebased and force pushed.

**Comment by lucaswoj** [comment]
Rad! Looks like it is working now.

File: debug/index.html (hunk 1 above)
Comment (lucaswoj) [inline debug/index.html]: Right now, this property literally controls the number of spawned workers. If we change the implementation in the future, we can change the property name 😄

Let's go with `workerCount` for an abbreviation-free, grammatically correct, and [best-practice compliant name](http://stackoverflow.com/questions/3742650/naming-conventions-for-number-of-foos-variables).

**Comment by anandthakker** [comment]
Sweet!

Did you see my [note](https://github.com/mapbox/mapbox-gl-js/pull/2655#discussion_r65649405) about the option name?  Easy to change, just lemme know.

**Comment by anandthakker** [comment]
Amended the commit:
 - `s/maxWorkers/workerCount/`
 - Add jsdoc for `workerCount` option

I believe this should now be ready

**Comment by anandthakker** [comment]
➡️ #2666

**Pull request closed without merging**